odl/space/pspace.py

```diff
@@ -1064,6 +1161,89 @@ def ufuncs(self):
         """
         return ProductSpaceUfuncs(self)
 
+    @property
+    def real(self):
+        """Real part of the element.
+
+        Examples
+        --------
+        >>> space = odl.ProductSpace(odl.cn(3), odl.cn(2))
+        >>> x = space.element([[1 + 1j, 2, 3 - 3j], [-1+2j, -2-3j]])
+        >>> x_real = x.real
+        >>> real_space = odl.ProductSpace(odl.rn(3), odl.rn(2))
+        >>> expected_result = real_space.element([[1, 2, 3], [-1, -2]])
+        >>> x_real == expected_result
+        True
+        """
+        real_part = [part.real for part in self.parts]
+        return self.space.real_space.element(real_part)
+
+    @real.setter
+    def real(self, newreal):
+        """Setter for the real part.
+
+        This method is invoked by ``x.real = other``.
+
+        Parameters
+        ----------
+        newreal : array-like or scalar
+            Values to be assigned to the real part of this element.
+
+        Examples
+        --------
+        >>> space = odl.ProductSpace(odl.cn(3), odl.cn(2))
+        >>> x = space.element([[1 + 1j, 2, 3 - 3j], [1 + 1j, 2]])
+        >>> zero = space.real_space.zero()
+        >>> x.real = zero
+        >>> x
+        space.element([[1j, 0j, -3j], [1j, 0]])
+
+        Other array-like types:
+
+        >>> x.real = 1.0
+        >>> x
+        space.element([[1 + 1j, 1, 1 - 3j], [1 + 1j, 1]])
+        >>> x.real = [[2, 3, 4], [5, 6]]
+        >>> x
+        space.element([[2 + 1j, 3, 4 - 3j], [5 + 1j, 6]])
+        """
+        if newreal in self.space.field:
+            for part in self.parts:
+                part.real.data[:] = newreal
+        elif newreal in self.space.real_space or newreal in self.space:
+            for part, newreal_subpart in zip(self.parts, newreal):
+                part.real.data[:] = newreal_subpart.real.data[:]
+        elif self.space.shape[0] == np.shape(newreal)[0]:
+            for part, newreal_subpart in zip(self.parts, newreal):
+                part.real.data[:] = np.real(newreal_subpart)
+        else:
+            raise ValueError('new real part does not the correct first '
+                             'dimension, expected {} but got {}'
+                             ''.format(self.space.shape[0],
+                                       np.shape(newreal)[0]))
```

Comment:
Overall the logic seems a bit too complicated for the cases it handles, and it misses out one case. Here are the cases that I think should be distinguished:

1. RHS not iterable (likely a scalar) -> call each `part.real` setter with this RHS (don't check `field` or so)
2. Space is a power space and RHS an element of the base space -> call each `part.real` setter with this RHS (i.e., the same as in case 1, can be merged)
   The reason for this (and that it comes before the next case) is that Numpy arrays behave the same: implicit new axes are added *to the left*, where leftmost is the power of the power space in our case:
   ```python
   >>> a = np.zeros((2, 2))
   >>> a.real = [1, 2]
   >>> a
   array([[ 1.,  2.],
          [ 1.,  2.]])
   ```
   **Note:** This also needs to include *element-like* objects for the base space.
3. `len(rhs) == len(self)` -> call `part[i].real` setter with `rhs[i]`, i.e., pretty much the "standard" case.
4. Error

So an implementation of that logic would be:
```python
try:
    iter(newreal)
except TypeError:
    for part in self.parts:
        part.real = newreal
    return

if self.is_power_space:
    try:
        newreal = self.space[0].element(newreal)
    except (ValueError, TypeError):  # maybe need to catch more, e.g. `LinearSpaceTypeError`
        pass
    else:
        for part in self.parts:
            part.real = newreal
elif len(newreal) == len(self):
    for part, new_re in zip(self.parts, newreal):
        part.real = new_re
else:
    raise ValueError('invalid')  # better error
```

Reply:
Your logic is much cleaner than mine, no doubt about that ;-) However, it passes all the current tests except when `newreal = [[0, 1 + 1j, 2], [3, 4]]`. Now, that a test for setting the real part crashes when the element we try to set it to is complex might seem logical. But what I do not like is that this is the ONLY test with complex values where it crashes. All other tests pass (with warnings of course). Should we care about this inconsistency, or just be happy that it indeed actually captures one error correctly and not just gives warnings in this case?

Reply:
Well, this implementation can have a quite laid-back attitude and just distribute work down in the way that makes most sense. That also means that raising errors is up to lower layers, so I wouldn't worry about that too much.
Regarding warnings vs. errors, we've been more and more going along with Numpy on most parts, simply because they probably have thought through a bunch of things and solved issues we don't even know about... and it's convenient! 😄 

Anyway, what is this one error in the test?

Reply:
This fails
```python
import odl
space = odl.ProductSpace(odl.uniform_discr(0, 1, 3, dtype=complex), odl.cn(2))
x = odl.util.testutils.noise_element(space)
newreal = [[0, 1 + 1j, 2], [3, 4]]
x.real = newreal
```
and gives
```
Traceback (most recent call last):

  File "<ipython-input-3-ede4021b455c>", line 5, in <module>
    x.real = newreal

  File "[...]/odl/space/pspace.py", line 1249, in real
    part.real = new_re

  File "[...]/odl/discr/lp_discr.py", line 615, in real
    self.tensor.real = newreal

  File "[...]/odl/space/npy_tensors.py", line 1251, in real
    self.real.data[:] = newreal

TypeError: can't convert complex to float
```
while for example the following passes
```python
import odl
space = odl.ProductSpace(odl.uniform_discr(0, 1, 3, dtype=complex), odl.cn(2))
x = odl.util.testutils.noise_element(space)
newreal = odl.util.testutils.noise_element(space)
x.real = newreal
```

Reply:
In fact, just chaining to a power space makes the first code snippet pass:
```
import odl
space = odl.ProductSpace(odl.uniform_discr(0, 1, 3, dtype=complex), 2)
x = odl.util.testutils.noise_element(space)
newreal = [[0, 1 + 1j, 2], [3, 4, 5]]
x.real = newreal
```

Reply:
For me all 3 versions work, which may be a matter of Numpy version. So my conclusion is that we should just let Numpy (or whatever backend we use) do what it does, and unless there's a fundamental difference we don't add new behavior. That also means less work for you 😉